src/main/java/org/testng/internal/ExitCodeListener.java

```diff
@@ -0,0 +1,106 @@
+package org.testng.internal;
+
+import org.testng.IReporter;
+import org.testng.ISuite;
+import org.testng.ISuiteResult;
+import org.testng.ITestContext;
+import org.testng.ITestListener;
+import org.testng.ITestResult;
+import org.testng.xml.XmlSuite;
+
+import java.util.BitSet;
+import java.util.List;
+
+public class ExitCodeListener implements ITestListener, IReporter {
```

Comment:
Is it possible to not implement a reporter for that and have only one or more listeners?

Reply:
Why do you ask? I chose the reporter path, because that would allow people to prune through their failed/skipped tests and remove the ones that shouldn't be there (typically done when working with a retry analyzer). In those cases, the status of the test would not 0, if we did this via any other listener.

Reply:
Because of #584. But I didn't realize the retry analyzer effect :/

Reply:
Ah ok! So you are saying that there can be a situation wherein a person might remove off the mandatory reporters. Understood. But this new listener doesn't care about the boolean `m_useDefaultListeners`. It gets added all the time.

Here's the snippet.
```java
private void initializeDefaultListeners() {
  this.m_exitCode = new org.testng.internal.ExitCodeListener();
  addListener((ITestNGListener) this.m_exitCode);
  if (m_useDefaultListeners) {
    addReporter(SuiteHTMLReporter.class);
    addReporter(Main.class);
    addReporter(FailedReporter.class);
    addReporter(XMLReporter.class);
    if (System.getProperty("oldTestngEmailableReporter") != null) {
      addReporter(EmailableReporter.class);
    } else if (System.getProperty("noEmailableReporter") == null) {
      addReporter(EmailableReporter2.class);
    }
    addReporter(JUnitReportReporter.class);
    if (m_verbose != null && m_verbose > 4) {
      addListener(new VerboseReporter("[TestNG] "));
    }
  }
}
```

Reply:
If we want to fix #584, one option is to be able to remove all reporters. If the core business code uses reporters, we will be stuck to go in that way.

Reply:
@juherr - IMO `ExitCodeListener` shouldn't fall into this list, because it doesn't cause of any delays. It only works on the sizes of passed/failed/skipped tests per `ITestContext`. But the other `IReporter` implementations can perhaps cause this.